### PR TITLE
quic: multi-endpoint and quicpacket

### PR DESCRIPTION
**Pull request by jasnell**
This accomplishes two important things:

1. Significant simplification of sendbuf and txbuf (by removing those). Adds `QuicPacket` to simplify serialization and flow of outbound packet data.

2. Adds `QuicEndpoint` and multi-endpoint support (not yet fully tested)

### CODE REVIEW

**Comment by jasnell** [comment]
@addaleax .. this is ready to review.

File: src/node_quic_default_application.cc [hunk 1]
```diff
@@ -143,12 +144,12 @@ bool DefaultApplication::SendStreamData(QuicStream* stream) {
 
   for (;;) {
     Debug(stream, "Starting packet serialization. Remaining? %d", remaining);
-    MallocedBuffer<uint8_t> dest(Session()->GetMaxPacketLength());
+    auto packet = CreateStreamDataPacket();
```
Comment (addaleax) [inline src/node_quic_default_application.cc]: It’s easier to read if the type is spelled out ;)

File: src/node_quic_socket.cc [hunk 2]
```diff
@@ -134,21 +134,109 @@ void JSQuicSocketListener::OnServerBusy(bool busy) {
...
 void JSQuicSocketListener::OnDestroy() {
   // Do nothing here.
 }
 
+QuicEndpoint::QuicEndpoint(
+    Environment* env,
+    Local<Object> wrap,
+    QuicSocket* listener,
+    Local<Object> udp_wrap) :
+    BaseObject(env, wrap),
```
Comment (addaleax) [inline src/node_quic_socket.cc]: mini style nit: `:` goes on the line of the first initializer

File: src/node_quic_socket.h [hunk 3]
```diff
@@ -88,12 +90,170 @@ class JSQuicSocketListener : public QuicSocketListener {
...
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackFieldWithSize("data", data_.length());
+    tracker->TrackFieldWithSize("label", strlen(diagnostic_label_));
+  }
+
+  SET_MEMORY_INFO_NAME(QuicPacket);
+  SET_SELF_SIZE(QuicPacket);
+
+ private:
+  MaybeStackBuffer<uint8_t, NGTCP2_MAX_PKTLEN_IPV4> data_;
```
Comment (addaleax) [inline src/node_quic_socket.h]: This means we always allocate the max packet length, even if the data packet is smaller (which I assume is the common case)… is that okay?

Comment (jasnell) [reply src/node_quic_socket.h]: Currently, it's not yet clear what the most common case will be. We need to do some testing and see what a reasonable default max is. We can adjust this in the future once we have a better idea

File: src/node_quic_socket.cc (hunk 2 above)
Comment (jasnell) [reply src/node_quic_socket.cc]: I use this pattern fairly extensively throughout. These can be updates later on in a separate PR

File: src/node_quic_default_application.cc (hunk 1 above)
Comment (jasnell) [reply src/node_quic_default_application.cc]: Can I totally disagree with that? ;-)

**Comment by jasnell** [comment]
Landed!

File: src/node_quic_default_application.cc (hunk 1 above)
Comment (addaleax) [reply src/node_quic_default_application.cc]: You can, but it’s hard to reason about the code locally here if the type isn’t obvious :)

**Pull request closed without merging**